From: Joao Martins <joao.m.martins@oracle.com>
Subject: mm-hugetlb_vmemmap-remap-head-page-to-newly-allocated-page-v4
Date: Thu, 10 Nov 2022 12:12:14 +0000

add smp_wmb() to ensure page contents are visible prior to PTE write

Link: https://lkml.kernel.org/r/20221110121214.6297-1-joao.m.martins@oracle.com
Signed-off-by: Joao Martins <joao.m.martins@oracle.com>
Reviewed-by: Muchun Song <songmuchun@bytedance.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 mm/hugetlb_vmemmap.c |    7 +++++++
 1 file changed, 7 insertions(+)

--- a/mm/hugetlb_vmemmap.c~mm-hugetlb_vmemmap-remap-head-page-to-newly-allocated-page-v4
+++ a/mm/hugetlb_vmemmap.c
@@ -246,6 +246,13 @@ static void vmemmap_remap_pte(pte_t *pte
 	if (unlikely(addr == walk->reuse_addr)) {
 		pgprot = PAGE_KERNEL;
 		list_del(&walk->reuse_page->lru);
+
+		/*
+		 * Makes sure that preceding stores to the page contents from
+		 * vmemmap_remap_free() become visible before the set_pte_at()
+		 * write.
+		 */
+		smp_wmb();
 	}
 
 	entry = mk_pte(walk->reuse_page, pgprot);
_
